lru.go

```diff
@@ -6,70 +6,94 @@
 package lru
 
 import (
-	"container/list"
+	"fmt"
 )
 
+// maxLRUSize is the largest we can fit in a buffer with a 32-bit unsigned offset
+const maxLRUSize = (1<<32 - 1)
+
 // LRU implements a least-recently-used cache, evicting items from the cache if they have not been accessed in a while.
 type LRU struct {
-	maxSize int
-	keys    *list.List
-	values  map[interface{}]cacheEntry
+	size     int
+	maxSize  int
+	buf      []cacheEntry
+	elements map[interface{}]uint32
+	root     *cacheEntry
 }
 
 type cacheEntry struct {
-	listElem *list.Element
-	value    interface{}
+	prev, next uint32
+	key        interface{}
+	value      interface{}
 }
 
 // Create a new LRU cache that will hold no more than the given number of items,
 func New(size int) *LRU {
-	return &LRU{
-		maxSize: size,
-		keys:    list.New(),
-		values:  make(map[interface{}]cacheEntry, 0),
+	if size > maxLRUSize || size <= 0 {
+		panic("size must not be <= 0 or >= 2^32")
+	}
+	initialBufSize := size + 1
+	if initialBufSize > 100 {
+		initialBufSize = 101
 	}
+	lru := &LRU{
+		size:     0,
+		maxSize:  size,
+		buf:      make([]cacheEntry, initialBufSize),
+		elements: make(map[interface{}]uint32, initialBufSize),
+	}
+	lru.root = &lru.buf[0]
+	return lru
 }
 
 // Len gives the number of items in the cache
 func (lru *LRU) Len() int {
-	return len(lru.values)
+	return lru.size
 }
 
 // Add a new entry into the LRU cache
 func (lru *LRU) Add(key, value interface{}) {
-	entry, exists := lru.values[key]
+	elem, exists := lru.elements[key]
 	if exists {
-		lru.keys.MoveToFront(entry.listElem)
+		entry := &lru.buf[elem]
+		lru.moveToFront(elem, entry)
 		// Update the value
 		entry.value = value
 	} else {
```

Comment:
The bench tests seem to indicate that the common use case looks like:

```golang
_, ok := cache.Get(K)
if !ok {
  cache.Put(K, V)
}
```

With that in mind it may be beneficial (could reduce cache misses due to branching) to invert the block so the most common path is evaluated first.

Reply:
the benchmarks may be a bit skewed because I was explicitly testing Intern vs other uses of an LRU cache. Intern specifically inlines the 2 calls so that it returns the value if it exists or the final result, because that avoids doing double lookups.
The use case for LRU itself can be seen somewhere like here:
https://github.com/juju/txn/blob/90cbb53292a84c397765581432d543d6faa404a4/incrementalprune.go#L385

Where it is:
  * check Get(), if it succeeds, use it
  * if it fails, go do some expensive operation (in this case queue up the document to be read from the database)
  * then populate the cache with the result of the expensive operation.

You can also see in that code extensive use of the string cache for attributes of the objects. (The transaction ObjectId, and the txn_tokens which is an object_id in hex + '_' + nonce.) A lot of those values reference each other and thus are used over and over, so it reduces memory overhead, (and slightly speeds up comparison operations), if you share the strings instead of comparing new strings all the time. (comparison is faster because 'string' in go is a pointer to a length and a buffer, and you can do 'a.buffer == b.buffer' rather than strncmp(a.buffer, b.buffer).

Anyway, I would say the truly common case is that Get() gets a hit, because otherwise it isn't worth using a Cache. So we *should* optimize for the case where LRU is worth having and thus the hit rates are high. (Because if it wasn't, then you probably shouldn't put a cache there :).